kratos/spatial_containers/search_wrapper.h

```diff
@@ -0,0 +1,1178 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Vicente Mataix Ferrandiz
+//
+
+#pragma once
+
+// System includes
+#include <numeric>
+
+// External includes
+
+// Project includes
+#include "includes/geometrical_object.h"
+#include "includes/kratos_parameters.h"
+#include "utilities/timer.h"
+#include "utilities/search_utilities.h"
+#include "utilities/parallel_utilities.h"
+#include "spatial_containers/geometrical_objects_bins.h"
+#include "spatial_containers/spatial_containers.h"
+#include "spatial_containers/spatial_search_result_container.h"
+#include "spatial_containers/spatial_search_result_container_vector.h"
+
+namespace Kratos
+{
+///@addtogroup KratosCore
+///@{
+
+///@}
+///@name Kratos Classes
+///@{
+
+/**
+ * @class SearchWrapper
+ * @ingroup KratosCore
+ * @brief This is a search wrapper ready for MPI searches
+ * @details Must be adapted and specialized for every search object
+ * @author Vicente Mataix Ferrandiz
+ * @tparam TSearchObject The seach object considered
+ * @tparam TSpatialSearchCommunication The communication type to be used
+ */
+template<class TSearchObject, SpatialSearchCommunication TSpatialSearchCommunication = SpatialSearchCommunication::SYNCHRONOUS_HOMOGENEOUS>
+class KRATOS_API(KRATOS_CORE) SearchWrapper
```

Comment:
It is the same class for serial parallel, what changes is the data communicator. If we define two different classes for serial/parallel it would be more troublesome and difficult to use.